README.md
# Suspenders [![Build Status](https://secure.travis-ci.org/thoughtbot/suspenders.png?branch=master)](http://travis-ci.org/thoughtbot/suspenders)

Suspenders is the base Rails application used at [thoughtbot](http://thoughtbot.com/community).

  ![Suspenders boy](http://media.tumblr.com/1TEAMALpseh5xzf0Jt6bcwSMo1_400.png)

Installation
------------

First install the suspenders gem:

    gem install suspenders

Then run:

    suspenders projectname

This will create a Rails app in `projectname` using the latest version of Rails.

By default this script creates a new git repository. See below if you
want to use it against an existing repo.

Gemfile
-------

To see the latest and greatest gems, look at Suspenders'
[Gemfile](templates/Gemfile.erb), which will be appended to the default
generated projectname/Gemfile.

It includes application gems like:

* [Airbrake](https://github.com/airbrake/airbrake) for exception notification
* [Bourbon](https://github.com/thoughtbot/bourbon) for Sass mixins
* [Delayed Job](https://github.com/collectiveidea/delayed_job) for background
  processing
* [Email Validator](https://github.com/balexand/email_validator) for email
  validation
* [Flutie](https://github.com/thoughtbot/flutie) for and `body_class` view
  helper
* [High Voltage](https://github.com/thoughtbot/high_voltage) for static pages
* [jQuery Rails](https://github.com/rails/jquery-rails) for jQuery
* [Neat](https://github.com/thoughtbot/neat) for semantic grids
* [Postgres](https://github.com/ged/ruby-pg) for access to the Postgres database
* [Rack Timeout](https://github.com/kch/rack-timeout) to abort requests that are
  taking too long
* [Recipient Interceptor](https://github.com/croaky/recipient_interceptor) to
  avoid accidentally sending emails to real people from staging
* [Simple Form](https://github.com/plataformatec/simple_form) for form markup
  and style
* [Unicorn](https://github.com/defunkt/unicorn) to serve HTTP requests
* [Title](https://github.com/calebthompson/title) for storing titles in
  translations

And gems only for staging and production like:

* [New Relic RPM](https://github.com/newrelic/rpm) for monitoring performance
* [Rails 12 Factor](https://github.com/heroku/rails_12factor) to make running
  Rails 4 apps easier on Heroku

And development gems like:

* [Dotenv](https://github.com/bkeepers/dotenv) for loading environment variables
* [Pry Rails](https://github.com/rweng/pry-rails) for debugging
* [Spring](https://github.com/rails/spring) for fast Rails actions via
  pre-loading

And testing gems like:

* [Capybara](https://github.com/jnicklas/capybara) and
  [Capybara Webkit](https://github.com/thoughtbot/capybara-webkit) for
  integration testing
* [Factory Girl](https://github.com/thoughtbot/factory_girl) for test data
* [Formulaic](https://github.com/thoughtbot/formulaic) for integration testing
  HTML forms
* [RSpec](https://github.com/rspec/rspec) for unit testing
* [RSpec Mocks](https://github.com/rspec/rspec-mocks) for stubbing and spying
* [Shoulda Matchers](https://github.com/thoughtbot/shoulda-matchers) for common
  RSpec matchers

Other goodies
-------------

Suspenders also comes with:

* The [`./bin/setup`][bin] convention for new developer setup
* Rails' flashes set up and in application layout
* A few nice time formats set up for localization
* `Rack::Deflater` to [compress responses with Gzip][compress]
* [Fast-failing factories][fast]
* A [low database connection pool limit][pool]
* [Safe binstubs][binstub]
* [t() and l() in specs without prefixing with I18n][i18n]
* An automatically-created `SECRET_KEY_BASE` environment variable in all
  environments.
* Configuration for [Travis Pro][travis] continuous integration.

[bin]: http://robots.thoughtbot.com/bin-setup
[compress]: http://robots.thoughtbot.com/content-compression-with-rack-deflater/
[fast]: http://robots.thoughtbot.com/testing-your-factories-first
[pool]: https://devcenter.heroku.com/articles/concurrency-and-database-connections
[binstub]: https://github.com/thoughtbot/suspenders/pull/282
[i18n]: https://github.com/thoughtbot/suspenders/pull/304
[travis]: http://docs.travis-ci.com/user/travis-pro/

Heroku
------

You can optionally create Heroku staging and production apps:

    suspenders app --heroku true

This:

* Creates a staging and production Heroku app
* Sets them as `staging` and `production` Git remotes
* Configures staging with `RACK_ENV` and `RAILS_ENV` environment variables set
  to `staging`

Git
---

This will initialize a new git repository for your Rails app. You can
bypass this with the `--skip-git` option:

    suspenders app --skip-git true

GitHub
------

You can optionally create a GitHub repository for the suspended Rails app. It
requires that you have [Hub](https://github.com/github/hub) on your system:

    curl http://hub.github.com/standalone -sLo ~/bin/hub && chmod +x ~/bin/hub
    suspenders app --github organization/project

This has the same effect as running:

    hub create organization/project

Dependencies
------------

Suspenders requires the latest version of Ruby.

Some gems included in Suspenders have native extensions. You should have GCC
installed on your machine before generating an app with Suspenders.

Use [OS X GCC Installer](https://github.com/kennethreitz/osx-gcc-installer/) for
Snow Leopard (OS X 10.6).

Use [Command Line Tools for XCode](https://developer.apple.com/downloads/index.action)
for Lion (OS X 10.7) or Mountain Lion (OS X 10.8).

We use [Capybara Webkit](https://github.com/thoughtbot/capybara-webkit) for
full-stack JavaScript integration testing. It requires QT. Instructions for
installing QT are
[here](https://github.com/thoughtbot/capybara-webkit/wiki/Installing-Qt-and-compiling-capybara-webkit).

PostgreSQL needs to be installed and running for the `db:create` rake task.

Issues
------

If you have problems, please create a
[GitHub Issue](https://github.com/thoughtbot/suspenders/issues).

Contributing
------------

To update Suspenders' Ruby version, change `.ruby-version` and `.travis.yml`.

Please see CONTRIBUTING.md for further details.

Credits
-------

![thoughtbot](http://thoughtbot.com/images/tm/logo.png)

Suspenders is maintained and funded by
[thoughtbot, inc](http://thoughtbot.com/community).

The names and logos for thoughtbot are trademarks of thoughtbot, inc.

License
-------

Suspenders is Copyright © 2008-2014 thoughtbot. It is free software, and may be
redistributed under the terms specified in the LICENSE file.


NEWS.md
1.12.0 (May 23, 2014)

* Generate `config/newrelic.yml`
* Include [formulaic](https://github.com/thoughtbot/formulaic) gem for
integration testing
* Removed dependency on `timecop` gem. Calls to `Timecop.freeze` can be
replaced by calls to `travel_to`, added in `rails` 4.1. To learn about the new
helpers, read [this blog
post](http://brandonhilkert.com/blog/rails-4-1-travel-to-test-helper)

1.11.0 (May 17, 2014)

* Generate a Rails 4.1.1 app and implement fixes for compatibility.
* Escape ERb in secrets.yml
* Maintain ActiveRecord test schema.
* Make Shoulda Matchers work with Spring.
* Unify Ruby version for gem and suspended apps.
* Move SMTP variable settings out of initializer.
* Connect to Postgres on localhost.
* Add `bin/setup` for contributors.
* Improve and document TravisCI configuration.

1.10.2 (April 28, 2014)

* Fix bundling Bourbon and Neat.

1.10.1 (April 25, 2014)

* Fix bundling sass-rails.

1.10.0 (April 21, 2014)

* Generate a Rails 4.1 app.
* Generate a working .ruby-version for Ruby >= 2.1.0.
* Update Unicorn template to version now preferred by Heroku.

1.9.3 (April 13, 2014)

* Use FactoryGirl.lint instead of custom-generated factory-testing code.
* Fix Delayed::Job <-> Rails 4.1 dependency conflict.

1.9.2 (April 10, 2014)

* Join Heroku apps in bin/setup.
* Enable SMTP/TLS in SMTP settings.
* Silence an RSpec warning.

1.9.1 (April 7, 2014)

* Fix sass-rails environment NilClass error.

1.9.0 (March 24, 2014)

* Add `awesome_print` gem.
* Add `dev:prime` task placeholder for bootstrapping local dev data.
* Add fix for I18n deprecation warning from `enforce_available_locales`.
* Add generated `.travis.yml`.
* Remove `better_errors` because of issues with Unicorn.
* Remove fast-failing for RSpec; respect user's `~/.rspec` instead.
* Update New Relic agent.
* Update Rails to 4.0.3.

1.8.1 (February 19, 2014)

* Don't distribute rspec binstub with gem.

1.8.0 (February 18, 2014)

* Make the .git/safe directory in bin/setup.
* Require `rails_12factor` gem only on Heroku.
* Require mailer config on staging and production.
* Add rspec binstub.
* Fix .ruby-version on Ruby 2.1.0.
* Replace Flutie's `page_title` with `title` gem.
* Don't run factory specs twice.
* Inherit staging config from production.
* Internal: convert tests from Cucumber to RSpec.
* Don't include `prefilled_input.js`.
* Fix Rack class name - Deflater instead of Timeout.
* Add Pry Rails.
* Add Spring.
* Add Dotenv to development and test environments to load environment variables
  from the `.env` file.
* Reduce ActiveRecord connection pool from 5 to 2.

1.7.0 (December 6, 2013)

* Keep `db/schema.rb` under version control.
* Fast-fail if any part of `bin/setup` fails.
* Move secret key out of version control.
* Create `.ruby-version` in generated applications.
* Add placeholder modules and directories for feature specs.
* Improve README to include setup instructions.

1.6.0 (November 28, 2013)

* Do not create `.rspec` file as the settings are not project-specific.
* Generate RSpec binstub at `bin/rspec`.
* Fix stylesheet error on 500, 404, and 422 static pages.
* Add `--skip-git` option.
* Disable jQuery animations in Rails integration tests that execute JavaScript.
* Fix git remote bug.
* Add `Rack::Deflater` to compress responses with Gzip.

1.5.1 (September 10, 2013)

* Remove Turbolinks.
* Don't use Bundler's binstubs in `bin/setup`.
* Remove `--drb` now that we aren't using Spork.
* Set up DNS via Pow for development.
* Update gem versions.

1.5.0 (August 3, 2013)

* Add Neat.
* Replace Bourne with RSpec Mocks.
* Replace Sham Rack with WebMock.
* Remove dependency on `hub` gem.
* Clean up leftover Rails 3 conventions.

1.4.0 (July 21, 2013)

* Support Rails 4.
* Support Ruby 2.
* Remove jQuery UI.
* Factories spec works for non-ActiveRecord objects.
* Use New Relic RPM gem >= 3.5.7 for Heroku request queue accuracy.
* Abort RSpec runs on first failure.
* Replace custom email validator with gem.

1.3.0 (May 13, 2013)

* Switch web server from Thin to Unicorn
* Set up database before setting up RSpec so that the rspec:install task works
* Add Delayed::Job
* Clean up cruft from ActionMailer delivery configuration
* strong_parameters now raises an error in development
* Enforce Ruby 1.9.2+ in the gemspec

1.2.2 (March 14, 2013)

* Fix Syntax error in staging/production environment config files.
* Make Factory Girl available to development environment for generators and
  `rails console`.

1.2.1 (February 28, 2013)

* Use Ruby 1.9.3 and 2.0
* Update staging and production email delivery
* Remove Spork and Guard
* Add better_errors and binding_of_caller gems
* Fix ActiveRecord attributes' blacklist
* Add Flutie to Gemfile

1.2.0 (February 13, 2013)

* Upgrade Rails from 3.2.8 to 3.2.12 to keep pace with security patches.
* Improve staging environment on Heroku with staging `ENV` variables and
  overriding the recipient in staging email delivery.
* Remove Flutie, use Bourbon.
* Wrap all HTTP requests in a 5 second timeout.
* Don't use `attr_accessible` whitelists. Instead, configure Strong Parameters.
* Provide a `bin/setup` script.
* Force RSpec's `expect` syntax.
* Remove remaining references to Cucumber, complete RSpec + Capybara conversion.
* Improve Foreman/`.env`/`Procfile` interactions.

1.1.5 (October 22, 2012)

* Ignore `.env`.
* Link to thoughtbot/guides in generated README.
* Remove Cucumber in favor of RSpec + Capybara.
* Deliver emails in staging environment to an overriden email or set of emails.
* Encode database as UTF8.
* Bundle with binstubs using 37signals' directory convention.
* Configure time formats using localization.
* Add Ruby version to Gemfile.
* Add fast-failing spec that tests validity of factories.
* Use SimpleCov for C0 coverage.
* Configure RSpec with `--profile` flag to find slow-running specs.

1.1.4 (September 4, 2012)

* Always store UTC in the DB.
* Use Rails 3.2.8.

1.1.3 (August 7, 2012)


lib/suspenders/app_builder.rb
module Suspenders
  class AppBuilder < Rails::AppBuilder
    include Suspenders::Actions

    def readme
      template 'README.md.erb', 'README.md'
    end

    def raise_on_delivery_errors
      replace_in_file 'config/environments/development.rb',
        'raise_delivery_errors = false', 'raise_delivery_errors = true'
    end

    def raise_on_unpermitted_parameters
      action_on_unpermitted_parameters = <<-RUBY

  # Raise an ActionController::UnpermittedParameters exception when
  # a parameter is not explcitly permitted but is passed anyway.
  config.action_controller.action_on_unpermitted_parameters = :raise
      RUBY
      inject_into_file(
        "config/application.rb",
        action_on_unpermitted_parameters,
        before: "\nend"
      )
    end

    def provide_setup_script
      copy_file 'bin_setup', 'bin/setup'
      run 'chmod a+x bin/setup'
    end

    def provide_dev_prime_task
      copy_file 'development_seeds.rb', 'lib/tasks/development_seeds.rake'
    end

    def configure_generators
      config = <<-RUBY
    config.generators do |generate|
      generate.helper false
      generate.javascript_engine false
      generate.request_specs false
      generate.routing_specs false
      generate.stylesheets false
      generate.test_framework :rspec
      generate.view_specs false
    end

      RUBY

      inject_into_class 'config/application.rb', 'Application', config
    end

    def set_up_factory_girl_for_rspec
      copy_file 'factory_girl_rspec.rb', 'spec/support/factory_girl.rb'
    end

    def configure_newrelic
      template 'newrelic.yml.erb', 'config/newrelic.yml'
    end

    def configure_smtp
      copy_file 'smtp.rb', 'config/smtp.rb'

      prepend_file 'config/environments/production.rb',
        "require Rails.root.join('config/smtp')\n"

      config = <<-RUBY

  config.action_mailer.delivery_method = :smtp
  config.action_mailer.smtp_settings = SMTP_SETTINGS
      RUBY

      inject_into_file 'config/environments/production.rb', config,
        :after => 'config.action_mailer.raise_delivery_errors = false'
    end

    def enable_rack_deflater
      config = <<-RUBY

  # Enable deflate / gzip compression of controller-generated responses
  config.middleware.use Rack::Deflater
      RUBY

      inject_into_file 'config/environments/production.rb', config,
        :after => "config.serve_static_assets = false\n"
    end

    def setup_asset_host
      replace_in_file 'config/environments/production.rb',
        '# config.action_controller.asset_host = "http://assets.example.com"',
        "config.action_controller.asset_host = ENV.fetch('ASSET_HOST')"

      replace_in_file 'config/environments/production.rb',
        "config.assets.version = '1.0'",
        "config.assets.version = ENV.fetch('ASSETS_VERSION')"

      replace_in_file 'config/environments/production.rb',
        'config.serve_static_assets = false',
        'config.static_cache_control = "public, max-age=#{1.year.to_i}"'
    end

    def setup_staging_environment
      staging_file = 'config/environments/staging.rb'
      copy_file 'staging.rb', staging_file

      config = <<-RUBY

Rails.application.configure do
  # ...
end
      RUBY

      append_file staging_file, config
    end

    def setup_secret_token
      template 'secrets.yml', 'config/secrets.yml', force: true
    end

    def create_partials_directory
      empty_directory 'app/views/application'
    end

    def create_shared_flashes
      copy_file '_flashes.html.erb', 'app/views/application/_flashes.html.erb'
    end

    def create_shared_javascripts
      copy_file '_javascript.html.erb', 'app/views/application/_javascript.html.erb'
    end

    def create_application_layout
      template 'suspenders_layout.html.erb.erb',
        'app/views/layouts/application.html.erb',
        force: true
    end

    def remove_turbolinks
      replace_in_file 'app/assets/javascripts/application.js',
        /\/\/= require turbolinks\n/,
        ''
    end

    def use_postgres_config_template
      template 'postgresql_database.yml.erb', 'config/database.yml',
        force: true
    end

    def create_database
      bundle_command 'exec rake db:create db:migrate'
    end

    def replace_gemfile
      remove_file 'Gemfile'
      template 'Gemfile.erb', 'Gemfile'
    end

    def set_ruby_to_version_being_used
      create_file '.ruby-version', "#{Suspenders::RUBY_VERSION}\n"
    end

    def setup_heroku_specific_gems
      inject_into_file 'Gemfile', "\n\s\sgem 'rails_12factor'",
        after: /group :staging, :production do/
    end

    def enable_database_cleaner
      copy_file 'database_cleaner_rspec.rb', 'spec/support/database_cleaner.rb'
    end

    def configure_spec_support_features
      empty_directory_with_keep_file 'spec/features'
      empty_directory_with_keep_file 'spec/support/features'
    end

    def configure_rspec
      remove_file 'spec/spec_helper.rb'
      copy_file 'spec_helper.rb', 'spec/spec_helper.rb'
    end

    def configure_travis
      template 'travis.yml.erb', '.travis.yml'
    end

    def configure_i18n_in_specs
      copy_file 'i18n.rb', 'spec/support/i18n.rb'
    end

    def configure_background_jobs_for_rspec
      copy_file 'background_jobs_rspec.rb', 'spec/support/background_jobs.rb'
      run 'rails g delayed_job:active_record'
    end

    def configure_action_mailer_in_specs
      copy_file 'action_mailer.rb', 'spec/support/action_mailer.rb'
    end

    def configure_time_zone
      config = <<-RUBY
    config.active_record.default_timezone = :utc

      RUBY
      inject_into_class 'config/application.rb', 'Application', config
    end

    def configure_time_formats
      remove_file 'config/locales/en.yml'
      copy_file 'config_locales_en.yml', 'config/locales/en.yml'
    end

    def configure_rack_timeout
      copy_file 'rack_timeout.rb', 'config/initializers/rack_timeout.rb'
    end

    def configure_action_mailer
      action_mailer_host 'development', "#{app_name}.local"
      action_mailer_host 'test', 'www.example.com'
      action_mailer_host 'staging', "staging.#{app_name}.com"
      action_mailer_host 'production', "#{app_name}.com"
    end

    def fix_i18n_deprecation_warning
      config = <<-RUBY
    config.i18n.enforce_available_locales = true

      RUBY
      inject_into_class 'config/application.rb', 'Application', config
    end

    def generate_rspec
      generate 'rspec:install'
    end

    def configure_unicorn
      copy_file 'unicorn.rb', 'config/unicorn.rb'
    end

    def setup_foreman
      copy_file 'sample.env', '.sample.env'
      copy_file 'Procfile', 'Procfile'
    end

    def setup_stylesheets
      remove_file 'app/assets/stylesheets/application.css'
      copy_file 'application.css.scss',
        'app/assets/stylesheets/application.css.scss'
    end

    def gitignore_files
      remove_file '.gitignore'
      copy_file 'suspenders_gitignore', '.gitignore'
      [
        'app/views/pages',
        'spec/lib',
        'spec/controllers',
        'spec/helpers',
        'spec/support/matchers',
        'spec/support/mixins',
        'spec/support/shared_examples'
      ].each do |dir|
        run "mkdir #{dir}"
        run "touch #{dir}/.keep"
      end
    end

    def init_git
      run 'git init'
    end

    def create_heroku_apps
      path_addition = override_path_for_tests
      run "#{path_addition} heroku create #{app_name}-production --remote=production"
      run "#{path_addition} heroku create #{app_name}-staging --remote=staging"
      run "#{path_addition} heroku config:add RACK_ENV=staging RAILS_ENV=staging --remote=staging"
    end

    def set_heroku_remotes
      remotes = <<-RUBY

# Set up staging and production git remotes
git remote add staging git@heroku.com:#{app_name}-staging.git
heroku join --app #{app_name}-staging

git remote add production git@heroku.com:#{app_name}-production.git
heroku join --app #{app_name}-production
      RUBY

      append_file 'bin/setup', remotes
    end

    def set_heroku_rails_secrets
      path_addition = override_path_for_tests
      run "#{path_addition} heroku config:add SECRET_KEY_BASE=#{generate_secret} --remote=staging"
      run "#{path_addition} heroku config:add SECRET_KEY_BASE=#{generate_secret} --remote=production"
    end

    def create_github_repo(repo_name)
      path_addition = override_path_for_tests
      run "#{path_addition} hub create #{repo_name}"
    end

    def setup_segment_io
      copy_file '_analytics.html.erb',
        'app/views/application/_analytics.html.erb'
    end

    def copy_miscellaneous_files
      copy_file 'errors.rb', 'config/initializers/errors.rb'
    end

    def customize_error_pages
      meta_tags =<<-EOS
  <meta charset='utf-8' />
  <meta name='ROBOTS' content='NOODP' />
      EOS

      %w(500 404 422).each do |page|
        inject_into_file "public/#{page}.html", meta_tags, :after => "<head>\n"
        replace_in_file "public/#{page}.html", /<!--.+-->\n/, ''
      end
    end

    def remove_routes_comment_lines
      replace_in_file 'config/routes.rb',
        /Rails\.application\.routes\.draw do.*end/m,
        "Rails.application.routes.draw do\nend"
    end

    def disable_xml_params
      copy_file 'disable_xml_params.rb', 'config/initializers/disable_xml_params.rb'
    end

    def setup_default_rake_task
      append_file 'Rakefile' do
        "task(:default).clear\ntask :default => [:spec]\n"
      end
    end

    private

    def override_path_for_tests
      if ENV['TESTING']
        support_bin = File.expand_path(File.join('..', '..', 'spec', 'fakes', 'bin'))
        "PATH=#{support_bin}:$PATH"
      end
    end

    def factories_spec_rake_task
      IO.read find_in_source_paths('factories_spec_rake_task.rb')
    end

    def generate_secret
      SecureRandom.hex(64)
    end
  end
end
